### PR TITLE
Add matrix plotting

### LINKED ISSUE #387: Merge new matrix plotting code
I have been using code here
https://github.com/neurodata/maggot_models/blob/master/src/visualization/matrix.py
for plotting matrices and sorting by various metadata in various ways. 

May also want to merge this with seaborn clustermap functionality - a lot of what I have was pulled from there, in fact, but I'm unsure what makes the most sense in terms of implementation. 

Regardless, would like the following few features: 
- sorting a matrix rows and/or columns by categorical labels 
- sorting the categories themselves by some metadata 
- sorting items within categories by some metadata 
- coloring the borders of a matrix plot by some metadata 
- (possibly) showing a clustering of the rows or columns, like seaborn clustermap

### PR DESCRIPTION
**Pull request by zeou1**
#### Reference Issues/PRs

Resolves #387 

#### What does this implement/fix? Explain your changes.
Implements plotting matrix after sorting the matrix by some metadata:
- group a matrix by one or more categorical labels
- sort the groups by the size of the group or the mean of some other metadata of each group
- sort items within a group by some metadata
- allow multiple tick and color axis to categorize the data
- provides tutorial and test

#### Any other comments?
Here is a notebook to prove the effectiveness of the implementation: [notebook](https://nbviewer.jupyter.org/github/zeou1/graspy/blob/matrix-plotting/docs/tutorials/plotting/matrixplot.ipynb)

This implementation currently stands separately from the existing plot.py file, as I did not want to make accidental changes. If needed, I can merge my implementation into plot.py. The test for plot_matrix is also separated right now.

### CODE REVIEW

**Comment by zeou1** [comment]
Based on the output of the checks, it looks like I am failing all of the unit-and-doc-test because of an error in test_rdpg_corr.py . I've never made edits to this file, so I don't know if other people have encountered this issue before. As for the build-documentation check, it says that my Jupiter notebook is not part of any toctree yet. After digging around in the package, I found that the toctree is used to create html tutorial on the website, so should I attach my tutorial notebook to our official tutorial now or wait for review first?

**Comment by zeou1** [comment]
I commented out the last two lines of test_rdpg_corr.py, and was able to pass all the tests except for mypy. I think this PR is ready for review.

**Comment by PerifanosPrometheus** [comment]
**PR itself (on GitHub)**
- [x] PR has a descriptive and succinct title.
- [x] PR has a brief description of what was changed.
- [x] PR is addressing a change that has already been brought up in an issue. This will ensure that the change you propose is desired by the maintainers, and may provide a chance to discuss implementation details prior to PRing which will save everyone time.
- [ ] PR cites any related issues in the PR description and uses closing keywords for any issues that should be closed as a result of the PR.
- [x] PR does not have any extraneous file changes associated with it. One common example of an extraneous change is one that was already made in the target branch, but is showing up as part of this PR under Files changed. (**changes made to rdpg_corr due to bug in testing**)
 - [x] PR does not have any unresolved merge conflicts.
 - [ ] If implementing a major new feature or algorithm, a notebook demonstrating the proof of effectiveness is linked in the PR description. Depending on the feature, this kind of validation may take too long to run as a test or in a tutorial notebook.

**Style**

  - [ ] Code follows Python variable naming conventions, one description is here. Most variables should be snake_case and classes shoud be defined as UpperCamelCase, for example.
  - [ ] Code uses descriptive variable names throughout.
  - [ ] Code has no commented out "junk code."
  - [ ] Line lengths are short (recommended <88 chars). This includes docstrings. Note: black tries to shorten line lengths but sometimes it is unable to do so automatically, especially for docstrings.

**Documentation**

 - [x] Any new public classes/functions have docstrings.
 - [x] Any new public classes/functions have been added to the appropriate .rst file here to be rendered by Sphinx.
 - [x] Any major new features (like a new algorithm) are accompanied by a succinct tutorial notebook. Or, if it is agreed upon with the maintainers that this is out of scope for the current PR, a new issue has been created specifying that we need a tutorial notebook for this functionality.
 - [x] Any new tutorial notebooks have been added to the appropriate folder here and included in the .rst file here to be rendered by Sphinx.
 - [ ] Modifications to the documentation render appropriately in the Netlify build.
 - [ ] References to parameters in the documentation are wrapped in double backticks (`` on either side of the parameter).
 - [ ] References to classes/function/documentation outside of this package use intersphinx to automatically link to the relevant documentation pages.

**Testing**

 - [x] New public classes/functions are tested to ensure they achieve the desired output.
 - [x] New public classes/functions are tested to ensure proper errors are thrown when invalid inputs are passed.

**Comment by PerifanosPrometheus** [comment]
@zeou1 is there any Jupyter Notebook attached to the PR for validation?

**Comment by zeou1** [comment]
> @zeou1 is there any Jupyter Notebook attached to the PR for validation?
> 
> Edit: I am not sure whether you may have discussed with Pedigo and the tutorial Notebook is enough for validation in your case

@bdpedigo Do you think the tutorial notebook is enough for validation purposes?

**Comment by bdpedigo** [comment]
> > @zeou1 is there any Jupyter Notebook attached to the PR for validation?
> > Edit: I am not sure whether you may have discussed with Pedigo and the tutorial Notebook is enough for validation in your case
> 
> @bdpedigo Do you think the tutorial notebook is enough for validation purposes?

yes. i havent looked at it. but as long as it demonstrates all of the features we say we support

**Comment by PerifanosPrometheus** [comment]
@bdpedigo I think this PR is ready for review

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:676]: make sure you remove any "junk code" prior to getting a review

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:483]: not sure why you took out the possibility for separate row/col information? can you explain?

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:676]: Okay, that's my bad, I though I got rid of all of them

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:483]: I can add it back if that's a necessary feature, but I just removed it because I thought it was unnecessary and would be inconvenient to use the function

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:483]: it is a necessary feature.

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:483]: that was the logic behind having `adjplot` as well if one doesn't need those features.

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:483]: okay cool, I will do that then.

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:483]: Since I am doing separate row/col information, should I have documentation for each of the row/col information, or can I have one documentation per pair in matrixplot?

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:483]: refer to seaborn clustermap for an example of what it should look like

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:170]: can also be a dict?

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:343]: rename to scattermap

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:693]: make this into a private function or something since it is repeated below

**Review by bdpedigo — CHANGES_REQUESTED** [review]
will leave more thorough comments after you add the row_ col_ distinctions

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:174]: I changed the format description to string or list of strings for color_class and group_class, I don't know if that's accurate

**Comment by zeou1** [comment]
@bdpedigo I added the separate row/col back to matrixplot and fixed other issues too.

File: graspologic/plot/plot_matrix.py
Comment (tliu68) [inline graspologic/plot/plot_matrix.py:208]: Do you think it would be helpful to include a cbar for the classes sometimes?

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:208]: It would be helpful, but the cbar sometimes makes the figure only shows up partially.

File: graspologic/plot/plot_matrix.py
Comment (tliu68) [inline graspologic/plot/plot_matrix.py:746]: It looks to me that the ticks are on or off for all classes. What if I had several classes but I only wanted to show ticks for some of them (maybe because some classes had too many categories to show or I had too many classes, etc)? Do you think we should accommodate cases like that as well?

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:746]: Right now the ticks are drawn as separate axis for all the group_class classes. If the user wants to sort the matrix by two classes, then there will be two tick axes, each separating the categories in each class.

Comment (tliu68) [reply graspologic/plot/plot_matrix.py:746]: So it either shows both tick axes or does not show any?

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:746]: yes

**Comment by tliu68** [comment]
In the tutorial, under "The use of item_order", 2 cases (sorting or not sorting by item_order) were shown. I understand the difference between the 2 cases, but am I supposed to see any differences from the visuals? (just trying to figure out what message I'm supposed to get from the comparison between the 2 figs)

**Comment by zeou1** [comment]
> In the tutorial, under "The use of item_order", 2 cases (sorting or not sorting by item_order) were shown. I understand the difference between the 2 cases, but am I supposed to see any differences from the visuals? (just trying to figure out what message I'm supposed to get from the comparison between the 2 figs)

Yes, I agree that the visual for the item_order is not straight forward. If you look at the color axis, you can see that the color is in the same sequence in the second plot.

File: graspologic/plot/plot_matrix.py
Comment (tliu68) [reply graspologic/plot/plot_matrix.py:746]: Right so I was asking if you think it would be useful to have the option of showing some, but not necessarily all, of the tick axes

File: graspologic/plot/plot_matrix.py
Comment (tliu68) [reply graspologic/plot/plot_matrix.py:208]: Where were you trying to put the cbar?

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:208]: I didn't regulate the position of the cbar, and it would usually show up on the right side. I usually see partial images in VS Code, so it could be a problem with VS Code rather than the program itself.

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:746]: Yes I can do that. If users don't specify, then all tick axes will be shown, and if they do, then only the particular one will.

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:174]: can palette be a dict? I think we want this capability

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:492]: should we allow for multiple? as in if color_class is a list, this can be a list of the same length

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:488]: @bdpedigo reminder to think about parameter names here

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:488]: should we change `col_group_class` to `col_group`?

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:488]: seaborn calls the colors just `row_colors`, should we do that? Also, most other seaborn funcs use `hue` instead

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:746]: @tliu68 isn't this what already happens if you make item_order a categorical variable? Please let me know if I'm not understanding. I do think we want this functionality but I think it is (mostly) taken care of by that.

Comment (tliu68) [reply graspologic/plot/plot_matrix.py:746]: Ah right! I agree!

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:174]: Yes, it turns out that palette can be a dict, I just forgot to include it in the documentation

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:502]: @zeou1 @tliu68 

I think we should either do
`row_group` and `row_group_order`
or
`row_class` and `row_class_order`
I have a slight preference for the first one ("class" can be an overloaded term in code/ML).
Thoughts? Any better ideas?

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:502]: i also think `row_color_class` -> `row_color`

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:502]: not sure about `col_item_order` -> `col_order`. I think having `item` in there is actually a little bit more clear.

**Comment by bdpedigo** [comment]
@zeou1 I don't see your tutorial notebook in the netlify render

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:545]: in my original implementation, these could also be passed as lists (like a list of labels) is that still possible?

**Comment by zeou1** [comment]
> @zeou1 I don't see your tutorial notebook in the netlify render

@bdpedigo That's weird because I included the notebook in tutorial.rst and it did show up in the past. But I am also getting a new error in build-documentation this time: 

/home/runner/work/graspologic/graspologic/docs/tutorials/plotting/matrixplot.ipynb:11:Definition list ends without a blank line; unexpected unindent.

This might have something to do with the issue, but I can't quite find the problem.

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:545]: Do you mean a list of strings or something else?

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [inline graspologic/plot/plot_matrix.py:88]: @zeou1 all of this code is to let the user pass arguments for anything that are not references to `meta` but rather just lists or arrays or any iterable. as in, a list of "group_class" labels, for example. does that make sense?

Comment (zeou1) [reply graspologic/plot/plot_matrix.py:88]: Yes, I understanded it, but I didn't know what would be the right data type for these

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:88]: array-like of the same shape as the corresponding matrix axis. 

somewhere in there, I also allow for list of list, which I'm not sure if we'd actually want to support. we should decide and make it clear what we support.

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:502]: Yes, I agree with @bdpedigo that we should do `row_group` and `row_group_order`. And maybe we can do `row_color_group`, because I feel like `row_color` could also mean the color of the axis rather than the classifying group of the axis. And I agree we should use`col_item_order`.

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:502]: part of my hesitation with "group" at the end of color is that the colors aren't necessarily groupings, they can be continuous. and even if they are categorical, they are not necessarily "grouped". see what I mean?

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:88]: this is in response to https://github.com/microsoft/graspologic/pull/546#discussion_r537874963

**Comment by bdpedigo** [comment]
> > @zeou1 I don't see your tutorial notebook in the netlify render
> 
> @bdpedigo That's weird because I included the notebook in tutorial.rst and it did show up in the past. But I am also getting a new error in build-documentation this time:
> 
> /home/runner/work/graspologic/graspologic/docs/tutorials/plotting/matrixplot.ipynb:11:Definition list ends without a blank line; unexpected unindent.
> 
> This might have something to do with the issue, but I can't quite find the problem.

sounds like a bug somewhere in the notebook, you can always try converting to python file in VS code and seeing if black or a linter find any errors. not sure what else to advise, but will need to be fixed prior to merging.

**Comment by bdpedigo** [comment]
also note that you have merge conflicts that will also need to be addressed

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:222]: i see the utility of this, but I think we should open this as a separate issue. would need to figure out how to deal with many different possible color maps if many different colors are present. also, setting `cbar` to True would look bad, we'll need to do it ourselves on a custom axis.

File: graspologic/plot/plot_matrix.py
Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:545]: one other possible input is "array-like". more or less just means {list, np.ndarray, pd.Series}

**Comment by bdpedigo** [comment]
- changing the name of a few main arguments 
   - updating these in the tutorials, docs as well
- making sure the documentation is clear about the possibility of passing in array-likes of metadata, as opposed to the dataframe and references to it
- adding an example of the above in the tutorial and making sure this functionality works as expected
- small comment about the outer ticks looking different visually

**Comment by zeou1** [comment]
@bdpedigo I've made all the updates we talked about and are passing almost all the tests, but somehow for the last couple of pushes, I've been getting a different unit-and-doc-test every time

**Comment by zeou1** [comment]
@bdpedigo would you like to review this PR sometime early this week?

**Review by bdpedigo — CHANGES_REQUESTED** [review]
- tutorial notebooks need to be cleared of output prior to committing. this will force netlify/read the docs to actually build them each time. I did this for your tutorial, but we'll see if it completes running in time. Sometimes notebooks are too long for netlify to build them, in which case you'll have to make it shorter if you get that error.
- because this is a tutorial, we should make it as clear and concise as possible. for instance, I'm sure you can think of a faster/shorter way to define `wtargs` and `wt` in a couple of places. You also don't need to repeat them if they were defined previously. 
- i made a few minor changes to language in the tutorial
- High level comment: At the beginning of class I remember talking about how to implement this stuff at a very high level - for instance should there be classes that do a lot of the work like in seaborn. In this case, i think having classes to keep track of state behind the scenes would have been helpful - for instance, I would like the user to be able to control the `fontsize` in many places, which would be a huge pain as implemented because it would require passing that keyword argument through several layers of functions. so for the purposes of you finishing Sprint 1 now, we can leave as is and implemented as functions, but I think in the future a lot of the functionality could be refactored into a class behind the scenes and the code would end up simpler. happy to talk more after this sprint is done.

Otherwise, just a few minor comments attached to this review

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:536]: If it's only one word over, do I need to do a second line?

Comment (bdpedigo) [reply graspologic/plot/plot_matrix.py:536]: I guess you can leave just be aware for future docstrings

File: graspologic/plot/plot_matrix.py
Comment (zeou1) [reply graspologic/plot/plot_matrix.py:289]: Yes, I think you are right. This part of the code was inherited from your implementation, but I just ran some tests and saw no difference with or without back separators.

**Comment by zeou1** [comment]
> * tutorial notebooks need to be cleared of output prior to committing. this will force netlify/read the docs to actually build them each time. I did this for your tutorial, but we'll see if it completes running in time. Sometimes notebooks are too long for netlify to build them, in which case you'll have to make it shorter if you get that error.
> * because this is a tutorial, we should make it as clear and concise as possible. for instance, I'm sure you can think of a faster/shorter way to define `wtargs` and `wt` in a couple of places. You also don't need to repeat them if they were defined previously.
> * i made a few minor changes to language in the tutorial
> * High level comment: At the beginning of class I remember talking about how to implement this stuff at a very high level - for instance should there be classes that do a lot of the work like in seaborn. In this case, i think having classes to keep track of state behind the scenes would have been helpful - for instance, I would like the user to be able to control the `fontsize` in many places, which would be a huge pain as implemented because it would require passing that keyword argument through several layers of functions. so for the purposes of you finishing Sprint 1 now, we can leave as is and implemented as functions, but I think in the future a lot of the functionality could be refactored into a class behind the scenes and the code would end up simpler. happy to talk more after this sprint is done.
> 
> Otherwise, just a few minor comments attached to this review

I used a shorter way to define `wtargs` and `wt`. I believe what you meant by repetitive definition is the last part where I show array_like sorting_kws, I had to redefine the matrix because it was randomly sorted.

**Comment by bdpedigo** [comment]
@zeou1 @ me again if done addressing comments

**Comment by zeou1** [comment]
@bdpedigo yes, all comments have been addressed except for the high-level one.

**Comment by bdpedigo** [comment]
@zeou1 somehow your commit here reverted all of the changes I made to the tutorial notebook https://github.com/microsoft/graspologic/pull/546/commits/f8a0e484145e469eba869c9012515fd06c062572

**Review by bdpedigo — APPROVED** [review]
approve, thanks @zeou1!

**Comment by zeou1** [comment]
> approve, thanks @zeou1!

Love it!

**Comment by zeou1** [comment]
@bdpedigo I found a few typos in the tutorial and fixed them, should not have a big impact on the implementation

**Comment by bdpedigo** [comment]
> @bdpedigo I found a few typos in the tutorial and fixed them, should not have a big impact on the implementation

appreciated!

**Comment by zeou1** [comment]
@bdpedigo is there still anything I can do to help this PR pass?

**Comment by bdpedigo** [comment]
Somehow I broke something in the docs and I haven't had time to fix. Will not affect your grade. Welcome to try to fix it

**Comment by zeou1** [comment]
> Somehow I broke something in the docs and I haven't had time to fix. Will not affect your grade. Welcome to try to fix it

I encountered the same issue once before, and it just went away after another push. I couldn't figure out what's wrong with the docstring either.

**Comment by zeou1** [comment]
I found an explanation for the error in build-reference-documentation here: https://developer.mantidproject.org/Standards/DocumentationGuideForDevs.html
But I don't understand why this error only shows up on this PR even though the reference has the same format as plot.py

**Comment by bdpedigo** [comment]
I don't see the error in the page you linked?

The docs are actually building (see this) 
https://603701b46c878f000782a7ba--graspologic.netlify.app/reference/reference/plotting.html#matrixplot

but I'm not sure why that error is still being thrown. I guess just cause we have that particular check set up to treat warnings as errors.

**Comment by zeou1** [comment]
It's this one: https://developer.mantidproject.org/Standards/DocumentationGuideForDevs.html#explicit-markup-ends-without-a-blank-line-unexpected-unindent

**Comment by zeou1** [comment]
I suspect it might have to do with the `meta` reference, but I don't have evidence

**Comment by bdpedigo** [comment]
i tracked it down to the list of sorting keywords so I just tried removing for now. also a few other small things I had to resolve because of a new way of hosting the docs so hopefully this works.

**Pull request merged**
